### ISSUE TITLE
incorrect Cisco AVpair handling?

### ISSUE THREAD

**Issue by stefanmdor** [inverse-inc/packetfence#935]
Hi,

I'm using the feature 'Role mapping by access list'. The access-list contains multiple lines, but only the first one is applied by the switch.
Packetfence Version: 5.3.1
Switch: WS-C2960X-24PS-L
Switch IOS Image: c2960x-universalk9-mz.152-3.E.bin

Debugging turns out that packetfence is sending the whole access-list as follows:
Oct  1 15:58:51.603: RADIUS:  Vendor, Cisco       [26]  39
Oct  1 15:58:51.603: RADIUS:   Cisco AVpair       [1]   33  "ip:inacl#101=permit tcp any any"
Oct  1 15:58:51.603: RADIUS:  Vendor, Cisco       [26]  39
Oct  1 15:58:51.603: RADIUS:   Cisco AVpair       [1]   33  "ip:inacl#101=permit udp any any"

The first entry with line number 101 is applied. The second entry has also line number 101 and is ignored by my switch. Other switches or other firmware might handle this differently. My resulting ACL is:
Extended IP access list GigabitEthernet1/0/1#v4#8A99BF4 (per-user)
    101 permit tcp any any

Is it possible to increment the line number for every ACL entry? I think, it has to look like this:
ip:inacl#101=permit tcp any any
ip:inacl#102=permit udp any any

kind regards
Stefan

**Comment by fdurand** [comment]
Hi Stefan,
try that:

```
diff --git a/lib/pf/Switch/Cisco/Catalyst_2960.pm b/lib/pf/Switch/Cisco/Catalyst_2960.pm
index 1344f84..b803780 100644
--- a/lib/pf/Switch/Cisco/Catalyst_2960.pm
+++ b/lib/pf/Switch/Cisco/Catalyst_2960.pm
@@ -506,8 +506,10 @@ sub returnRadiusAccessAccept {
         if( defined($user_role) && $user_role ne ""){
             my $access_list = $self->getAccessListByName($user_role);
             my @av_pairs;
+            my $acl_num = 101;
             while($access_list =~ /([^\n]+)\n?/g){
-                push(@av_pairs, $self->returnAccessListAttribute."=".$1);
+                push(@av_pairs, $self->returnAccessListAttribute($acl_num)."=".$1);
+                $acl_num ++;
                 $logger->info("[$mac] (".$self->{'_id'}.") Adding access list : $1 to the RA
             } 
             $radius_reply_ref->{'Cisco-AVPair'} = \@av_pairs; 
@@ -541,8 +543,8 @@ Returns the attribute to use when pushing an ACL using RADIUS
 =cut

 sub returnAccessListAttribute {
-    my ($this) = @_;
-    return "ip:inacl#101";
+    my ($this, $acl_num) = @_;
+    return "ip:inacl#$acl_num";
 }

 sub disableMABByIfIndex {
```

**Comment by stefanmdor** [comment]
Hi fdurand,

thanks for the quick reply. Unfortunately I'm not a developer and don't know how to use diff. :-(

If I paste your code within the shell, there are a lot of error message starting with:
diff: Unknown Option »--git«

Could you send me a patched file instead?

best regards
Stefan

**Comment by fdurand** [comment]
Hi Stefan,
save the code in a file in /usr/local/pf/ , like acl.diff then run:
patch -p1 --dry-run < acl.diff
if there is no error then
patch -p1 < acl.diff

And restart packetfence

**Comment by stefanmdor** [comment]
Hi, I get the following error:

patching file lib/pf/Switch/Cisco/Catalyst_2960.pm
patch: ***\* malformed patch at line 6: if( defined($user_role) && $user_role ne "") {

**Comment by fdurand** [comment]
```
diff --git a/lib/pf/Switch/Cisco/Catalyst_2960.pm b/lib/pf/Switch/Cisco/Catalyst_2960.pm
index 1344f84..b803780 100644
--- a/lib/pf/Switch/Cisco/Catalyst_2960.pm
+++ b/lib/pf/Switch/Cisco/Catalyst_2960.pm
@@ -506,8 +506,10 @@ sub returnRadiusAccessAccept {
         if( defined($user_role) && $user_role ne ""){
             my $access_list = $self->getAccessListByName($user_role);
             my @av_pairs;
+            my $acl_num = 101;
             while($access_list =~ /([^\n]+)\n?/g){
-                push(@av_pairs, $self->returnAccessListAttribute."=".$1);
+                push(@av_pairs, $self->returnAccessListAttribute($acl_num)."=".$1);
+                $acl_num ++;
                 $logger->info("[$mac] (".$self->{'_id'}.") Adding access list : $1 to the RADIUS reply");
             } 
             $radius_reply_ref->{'Cisco-AVPair'} = \@av_pairs; 
@@ -541,8 +543,8 @@ Returns the attribute to use when pushing an ACL using RADIUS
 =cut

 sub returnAccessListAttribute {
-    my ($this) = @_;
-    return "ip:inacl#101";
+    my ($this, $acl_num) = @_;
+    return "ip:inacl#$acl_num";
 }

 sub disableMABByIfIndex {
```

**Comment by stefanmdor** [comment]
Thanks a lot! Now it is working as expected. The resulting ACL on the switch now is:

Extended IP access list GigabitEthernet1/0/1#v4#8A99BF4 (per-user)
    101 permit tcp any any
    102 permit udp any any

Could you provide a patch for the file 'Catalyst_2960G.pm' too, please?

best regards
Stefan

**Comment by fdurand** [comment]
It should work since the 2960G module use 2960 module

**Comment by julsemaan** [comment]
@fdurand can you put this in devel and we can then close this issue

**Comment by fdurand** [comment]
Let me test it first on different Cisco switches.

**Comment by julsemaan** [comment]
This was fixed via #1151
